Add Cilium debugger images and default debugging configuration for kind, vscode

joestringer opened the pull request:
Review commit by commit.

A new `make kind-image-debug` target wraps the Cilium image with a dlv debugger wrapper,
allowing live debug of the container image after you deploy it.

New make targets introduced:
- `kind-debug-agent` - Bootstrap kind, install cilium, build the cilium-agent and run it
- `kind-image-agent-debug` - Build cilium-agent image with a debugger wrapper that pauses on startup. Loads it into kind but does not restart the pods.
- `kind-install-cilium` - Install cilium into the kind cluster, pointing to locally built kind images.

Usage:
* `make kind-debug-agent`
* attach the debugger to the target node on port 2345 and proceed
  * If you are using vscode then the .vscode/launch.json will automatically provide attach targets in your vscode debugger UI. Just pick your node target:
    ![image](https://user-images.githubusercontent.com/1243336/186947762-427b97fe-ea7c-4cd1-898f-fecd88cd5799.png)

Note that the Cilium pods will pause immediately on startup until you
have connected the debugger.

This PR exposes the debugger port for cilium-agent on each kind node as well, which is enabled by loading an image using `make kind-image-debug`.

Port allocations:
- 23401: First control plane node
- 2340*: Subsequent kind-control-plane nodes (if defined)
- 23411: First kind-worker node
- 2341*: Subsequent kind-worker nodes (if defined)

Delve should be listening on this port with multiprotocol support, so any IDEs or debugger frontends that understand either DAP or delve API v2 should be compatible with this approach.

joestringer committed:
Makefile: Add kind-image-debug target

This target wraps the default image with a dlv debugger wrapper,
allowing live debug of the container image after you deploy it.

Usage:
* make kind-image-debug
* restart target Cilium pods
* attach the debugger to the target node on port 2345 and proceed

Note that the Cilium pods will pause immediately on startup until you
have connected the debugger.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
contrib: Expose debugger port on each kind node

Expose the debugger port on each kind node, when enabled by loading an
image using `make kind-image-debug`.

Port allocations:
- 23401: First control plane node
- 2340*: Subsequent kind-control-plane nodes (if defined)
- 23411: First kind-worker node
- 2341*: Subsequent kind-worker nodes (if defined)

If you run the script with more than 9 control plane nodes or more than
9 worker nodes, then the port forwarding will be disabled. It's assumed
at that point that you will not be concurrently debugging all of those
nodes.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
.vscode: Add default workspace configurations

These configurations provide suggestions for which Visual Studio Code
extensions developers should install to be successful in their
development on Cilium, and provide default debugging targets for
connecting a debugger to kind clusters that are launched via:

    $ make kind
    $ # Install Cilium and point it to localhost:5000/cilium/cilium-dev:local
    $ make kind-image-debug
    $ # Restart individual Cilium pods

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
images: update cilium-{runtime,builder}

Signed-off-by: Joe Stringer <joe@cilium.io>

ldelossa commented on .vscode/launch.json:
```diff
@@ -0,0 +1,47 @@
+			"type": "go",
+			"request": "attach",
+			"mode": "remote",
+			"remotePath": "",
```
These and the similar are redundant and can be removed. Setting subsitutePath is the alternative, which ware doing.

joestringer replied:
`These` meaning just `remotePath`? I can try dropping that. I'm guessing type/request/mode are all still required?

ldelossa replied:
Yup "these" are "remotePath" values. Shouldnt be necessary

ldelossa commented on images/cilium/Dockerfile:
```diff
@@ -93,3 +93,20 @@ WORKDIR /home/cilium
+ARG TARGETARCH
+COPY --from=builder /go/bin/dlv /usr/bin/dlv
+RUN mv /usr/bin/cilium-agent /usr/bin/cilium-agent-bin
+COPY images/scripts/debug-wrapper.sh /usr/bin/cilium-agent
```
You technically don't need to do this whole switch-a-roo 

You can also just make the wrapper script look like this:

```
set -o errexit
set -o nounset
set -o pipefail

/usr/bin/dlv \
    --listen=:2345 \
    --log=true \
    --log-output=debugger,debuglineerr,gdbwire,lldbout,rpc \
    --accept-multiclient \
    --api-version=2 \
    dap
```

And the launch.json look like:

```
    "configurations": [
		{
			"name": "Attach to kind-control-plane-1",
			"type": "go",
			"request": "launch",
			"mode": "exec",
			"program": "/usr/bin/cilium-agent",
			"port": 23401,
			"host":"127.0.0.1",
			"showLog": true,
			"trace": "log",
			"logOutput": "rpc",
			"debugAdapter": "dlv-dap",
			"substitutePath": [{"from": "${workspaceFolder}", "to": "/go/src/github.com/cilium/cilium"}],
		},
```

The difference here is that delve is started as a dap server, but doesn't do anything until the client connects. When it does, the client tells Delve to launch cilium-agent. 

This **would** be beneficial, since you could ask delve to "rerun" and it could restart the binary, but unfortunately I think that is currently broken in Delve anyway. See: https://github.com/go-delve/delve/tree/master/Documentation/api/dap#single-client-mode

So i'll let you judge which route is better.

joestringer replied:
> The difference here is that delve is started as a dap server, but doesn't do anything until the client connects. When it does, the client tells Delve to launch cilium-agent.

:thinking: that's exactly the behaviour I was observing. If I follow right, what I do is move the real binary out of the way, move cilium-agent into the place where the regular binary would be, then when kubelet runs the pod, it just runs the $thing at /usr/bin/cilium-agent which _happens_ to be the wrapper script. Since it's the wrapper script, that will start the debugger servers (in this case, both DAP and Delve API v2), then it will start running the agent with standard args but pause it before any logic runs. With the current approach, I'm not modifying the Pod / DaemonSet YAMLs at all, they're stock standard. Incidentally, the wrapper script should also work for cilium-operator in the same way since there's no direct references to the binary.

I *think* you're suggesting to modify the pod YAML so that the entrypoint to the container is the path to this debugger script, which would also work. How do we customize the args to hand to dlv in this case? One thing I want to make sure happens is that we pass the same config flags as what was already in the pod, so that the configuration is properly implemented (ie it's not like just by running the debugger you randomly change settings that could impact reproducibility of some bug).

> This would be beneficial, since you could ask delve to "rerun" and it could restart the binary, but unfortunately I think that is currently broken in Delve anyway. 

I was doing this with this approach (clicking the button to restart the app), it seemed to work already :sweat_smile:  But maybe it doesn't work 100% the same as what you were saying - I think I noticed that the error counts would increase in the `kubectl get pods` output. So maybe I was just observing that the binary would exit with some error code when I click the restart button, then maybe I was relying on kubelet to restart the application (which it will happily do, though after a couple of times it may introduce CrashLoopBackoff timeouts). Since the wrapper script intercepts startup and pauses the pod waiting for the debugger client before launching the application, this was functionally very similar to what you're suggesting.

joestringer replied:
^^ Yeah confirmed this, I've been relying on k8s to restart the app, so if you do this too many times it starts to introduce arbitrary delays because it thinks that the app is failing too much. It's not a huge deal unless you're needing to constantly restart the agent to debug init logic. Not sure what's the best path to mitigate that at this point. Part of me wants to adopt your suggestion, I'm just not sure about a clean way to introduce this into the DaemonSet to keep the changes minimal vs. the real production DaemonSet.

ldelossa replied:
You know, now that you mention it, it would be nicer that we don't need to modify the daemonset at all. 

Just to be clear, I wasn't "suggesting" the above, just showing another way of going about it, in which the cilium-agent/cilium-agent-bin swap around wasn't strictly necessary. 

> ^^ Yeah confirmed this, I've been relying on k8s to restart the app, so if you do this too many times it starts to introduce arbitrary delays because it thinks that the app is failing too much. It's not a huge deal unless you're needing to constantly restart the agent to debug init logic. Not sure what's the best path to mitigate that at this point. Part of me wants to adopt your suggestion, I'm just not sure about a clean way to introduce this into the DaemonSet to keep the changes minimal vs. the real production DaemonSet.

Here's the kicker, I don't believe there is a way around this since Delve does not support the "Restart" DAP command. So we can't ask DAP to restart the application its been configured to launch. Instead, when it sees a disconnect it will just terminate the launched program and exit itself. 

It also appears there are no plans to support this: https://github.com/go-delve/delve/discussions/3049 (with a little comment from your's truly lolol).

Anyway, the above comments were just food for thought, but now that you explained why you went about it this way, and the Daemon set entry point considerations, I think the way we have it in this PR is probably the best.

aditighag reviewed (commented):
I forgot to share this feedback earlier, but there've been times when cilium agent would restart while I'm inspecting stuff when breakpoints are hit, and so I ended up disabling health probes altogether. I wonder if we should add a hidden helm option that'll automatically be set via .sh file. The file can build various debug images as well. 
Another thing is the ability to attach debugger to non-agent related binaries (e.g., cilium monitor, cni, etc). Although I've always done this manually as and when required.

ldelossa committed:
.nvim: Add debugger instructions and config

This commit adds a nvim-dap configuration at `.nvim/nvim-dap.lua` along
with instructions on how to get Neovim wired up and connected to our
debugging infrastructure.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

joestringer committed:
Makefile: Add kind-install target

This kind-install target is intended for developer usage. Notably, this
applies very high failure thresholds for the Cilium pods so that these
pods don't restart while you are actively testing / debugging.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
Makefile: Add kind-debug-agent make target

This target will provision kind, build agent debug images, and install
cilium into a cluster all in one.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
Makefile: Fix up kind-ready target

This target didn't previously actually check if there was a cluster
running, so it would pass even if there isn't a cluster. Fix it.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
Makefile: Templatize kind env variables

These variables will be used by a few targets, so create a template that
will provide access to the same vars from all relevant targets.

Signed-off-by: Joe Stringer <joe@cilium.io>

ldelossa commented on Makefile:
```diff
@@ -547,6 +547,41 @@ kind-image: ## Build cilium and operator images and import them into kind.
+		--chart-directory=$(ROOT_DIR)/install/kubernetes/cilium \
+		--helm-values=$(ROOT_DIR)/contrib/testing/kind-values.yaml \
+		--version= \
+		>/dev/null 2>/dev/null &
```
Will
```suggestion
  &>/dev/null &
```
suffice? It's not POSIX but its bash to YOLO

joestringer replied:
Is there a functional difference? I'm not so familiar with POSIX etc. so I'll be relying on mac developers to report if this target needs any further tweaks to work right.

joestringer replied:
Probably `>/dev/null 2>&1` is the standard way to do this:

https://unix.stackexchange.com/questions/590694/posix-compliant-way-to-redirect-stdout-and-stderr-to-a-file

joestringer replied:
```
$ git ap
diff --git a/Makefile b/Makefile
index 31f17c50b39b..20ca5eef98fc 100644
--- a/Makefile
+++ b/Makefile
@@ -560,7 +560,7 @@ kind-install-cilium: kind-ready ## Install a local Cilium version into the clust
                --chart-directory=$(ROOT_DIR)/install/kubernetes/cilium \
                --helm-values=$(ROOT_DIR)/contrib/testing/kind-values.yaml \
                --version= \
-               >/dev/null 2>/dev/null &
+               >/dev/null 2>&1 &
 
 .PHONY: kind-check-cilium
 kind-check-cilium:
```

ldelossa replied:
No functional difference, just less lines of code. I think this is a  bashism so any decently new version of bash will use it. 

@chancez may know if this is cool with Mac.

ldelossa replied:
Yup that works. You are a man of standards :-D

ldelossa commented on Makefile:
```diff
@@ -547,6 +547,41 @@ kind-image: ## Build cilium and operator images and import them into kind.
+		   kubectl get -n kube-system deploy/cilium-operator 2>/dev/null >/dev/null; then \
+		   break; \
+		fi; \
+		sleep 1; \
```
Can this 10 second wait loop be nixed for kubectl's `wait` command instead? https://www.mankier.com/1/kubectl-wait

Just removes a heuristic of a ten second wait.

joestringer replied:
Hmm. What I wanted to achieve with this is to make sure that the Cilium resources were installed into the cluster, so that when this returns, `kind-install` target can successfully update the images to point to the debug images.

Does `kind-install` need to set the images to point to the local/debug images? Maybe not. If we rely on this target always installing the local images via `$(ROOT_DIR)/contrib/testing/kind-values.yaml` then the images should be set correctly regardless of how `make kind-install` gets run, so then we can probably skip the wait here and also the image set commands in the `kind-install` target.

joestringer replied:
:bulb: also I just realized that I was assuming `kind-install` is the target to install _cilium_ into kind, but the make target sounds like it's installing kind. I can tweak that too. Maybe `kind-install-cilium` would be less confusing.

joestringer replied:
This will be a nice simplification. I'll test it out:

```
$ git diff HEAD~2
diff --git a/Makefile b/Makefile
index e644a80ce012..6a5b7d5c0d53 100644
--- a/Makefile
+++ b/Makefile
@@ -547,8 +547,8 @@ kind-image: ## Build cilium and operator images and import them into kind.
        $(MAKE) kind-image-agent
        $(MAKE) kind-image-operator
 
-.PHONY: kind-install-full
-kind-install-full:
+.PHONY: kind-install-cilium
+kind-install-cilium: kind-ready ## Install a local Cilium version into the cluster.
        @echo "  INSTALL cilium"
        # cilium-cli doesn't support idempotent installs, so we uninstall and
        # reinstall here. https://github.com/cilium/cilium-cli/issues/205
@@ -561,26 +561,11 @@ kind-install-full:
                --helm-values=$(ROOT_DIR)/contrib/testing/kind-values.yaml \
                --version= \
                >/dev/null 2>/dev/null &
-       for i in seq 1 10; do \
-               if kubectl get -n kube-system ds/cilium 2>/dev/null >/dev/null && \
-                  kubectl get -n kube-system deploy/cilium-operator 2>/dev/null >/dev/null; then \
-                  break; \
-               fi; \
-               sleep 1; \
-       done
-       if ! kubectl get -n kube-system ds/cilium 2>/dev/null >/dev/null || \
-          ! kubectl get -n kube-system deploy/cilium-operator 2>/dev/null >/dev/null; then \
-               echo "failed to install cilium" >&2; \
-               exit 1; \
-       fi;
-
-$(eval $(call KIND_ENV,kind-install))
-kind-install: kind-ready ## Install a local Cilium version into the cluster.
-       $(ECHO_CHECK) cilium is ready...
-       cilium status --wait --wait-duration 1s >/dev/null 2>/dev/null \
-               || $(MAKE) kind-install-full
-       kubectl -n kube-system set image daemonset/cilium cilium-agent=$(LOCAL_AGENT_IMAGE)
-       kubectl -n kube-system set image deployment/cilium-operator cilium-operator=$(LOCAL_OPERATOR_IMAGE)
+
+.PHONY: kind-check-cilium
+kind-check-cilium:
+       @echo "  CHECK  cilium is ready..."
+       cilium status --wait --wait-duration 1s >/dev/null 2>/dev/null
 
 # Template for kind debug targets. Parameters are:
 # $(1) agent target
@@ -604,7 +589,8 @@ kind-debug-agent: ## Create a local kind development environment with cilium-age
        # Not debugging cilium-operator here; any image is good enough.
        $(CONTAINER_ENGINE) push $(LOCAL_OPERATOR_IMAGE) \
                || $(MAKE) kind-image-operator
-       $(MAKE) kind-install
+       $(MAKE) kind-check-cilium \
+               || $(MAKE) kind-install-cilium
        @echo "Attach delve to localhost on these ports to continue:"
        @echo " - 23401: cilium-agent (kind-control-plane)"
        @echo " - 23411: cilium-agent (kind-worker)"
```

joestringer replied:
:heavy_check_mark: seems to do the trick both from fresh install and after running the target before.
